reference/best-practices: add Haproxy Best Practice (#1889)
pingcap/docs-cn: v3.0/reference/best-practices/haproxy-best-practices.md

![HAProxy](/media/haproxy.jpg)

### 安装 HAproxy

1. 使用 yum 安装 HAProxy。

   ```bash
   yum -y install haproxy
   ```

2. 验证 HAProxy 安装是否成功

   ```bash
   which haproxy
   ```

#### HAProxy 命令介绍

```bash
$ haproxy --help
HA-Proxy version 1.9.0 2018/12/19 - https://haproxy.org/
Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
```

|参数|描述|详情|
|:-----|:---|:---|
|-v|displays version|显示简略的版本信息|
|-vv|shows known build options|显示详细的版本信息|
|-d|enters debug mode|debug 模式开启|
| -db|only disables background mode.|仅禁止后台模式|
|-dM|[\<byte>] poisons memory with \<byte> (defaults to 0x50)|执行分配内存|
|-V|enters verbose mode (disables quiet mode)|启动过程显示配置和轮询信息
|-D|goes daemon|开启守护进程模式||
|-C|changes to \<dir> before loading files.||
|-W|master-worker mode.|主从模式|
|-q|quiet mode: don't display messages|静默模式，不输出信息|
|-c|check mode: only checks config files and exits|检查配置信息文件|
|-n|sets the maximum total # of connections (2000)|最大总连接数|
|-m|limits the usable amount of memory (in MB)|最大使用内存|

> |-N|sets the default, per-proxy maximum # of connections (2000)|单点最大连接数|

Reviewer: ```suggestion
|-N|sets the default, per-proxy maximum # of connections (2000)|设置单点最大连接数，默认为2000 |
```

|-L|sets local peer name (default to hostname)|本地实例对等名称|

> |-p|writes pids of all children to this file|HAProxy 进程 pid 信息写入 file |

Reviewer: ```suggestion
|-p|writes pids of all children to this file|将HAProxy 所有子进程的 PID 信息写入该文件 |
```

- [SSL](http://cbonte.github.io/haproxy-dconv/1.9/intro.html#3.3.2)：支持 HTTPS 通信和解析；
- [监控与统计](http://cbonte.github.io/haproxy-dconv/1.9/intro.html#3.3.3)：通过 web 页面可以实时监控服务状态以及具体的流量信息。

## 准备环境

在部署 HAProxy 之前，需准备好以下环境。

### 硬件要求

根据官方文档，对 HAProxy 的服务器硬件配置有以下建议，也可以根据负载均衡环境进行推算，在此基础上提高服务器配置。

|硬件资源|最低配置|
|---|---|
|CPU|2 cores, 3.5 GHz|
|Memory| 16 GB|
|Storage| 50 GB (SATA)|
|Network Interface Card| 万兆网卡|

### 依赖软件

根据官方文档，对操作系统和依赖包有以下建议，如果通过 yum 源部署安装 HAProxy 软件，依赖包无需单独安装。

### 操作系统

- Linux 2.4 on x86, x86_64, Alpha, Sparc, MIPS, PARISC
- Linux 2.6 / 3.x on x86, x86_64, ARM, Sparc, PPC64
- Solaris 8/9 on UltraSPARC 2 and 3
- Solaris 10 on Opteron and UltraSPARC
- FreeBSD 4.10 - 10 on x86
- OpenBSD 3.1 to -current on i386, amd64, macppc, alpha, sparc64 and VAX (check the ports)
- AIX 5.1 - 5.3 on Power™ architecture

### 依赖包

- epel-release
- gcc
- systemd-devel

## 部署 HAProxy

HAProxy 配置 Database 负载均衡场景操作简单，以下部署操作具有普遍性，不具有特殊性，建议根据实际场景，个性化配置相关的[配置文件](http://cbonte.github.io/haproxy-dconv/1.9/configuration.html)。

> **注意：**
>
> 官方目前建议使用 HAProxy 稳定版本 2.0，其特性参见 [2.0 特性](https://www.haproxy.com/blog/haproxy-2-0-and-beyond/)。

> ![HAProxy](/media/Haproxy.jpg)

Reviewer 2: ```suggestion
![HAProxy](/media/haproxy.jpg)
```
